src/libraries/System.Runtime/tests/System/ArrayTests.cs

```diff
@@ -4251,6 +4251,78 @@ public void Fill_InvalidStartIndexCount_ThrowsArgumentOutOfRangeException(int ar
             AssertExtensions.Throws<ArgumentOutOfRangeException>("count", () => Array.Fill(new string[arrayLength], "", startIndex, count));
         }
 
+        public class Bar
+        {
+            public string Value { get; set; }
+        }
+
+        public class Foo
+        {
+        }
+
+        private static Bar[] CreateBarArray()
+        {
+            return new Bar[]
+            {
+                new Bar() { Value = "0" },
+                new Bar() { Value = "1" },
+                new Bar() { Value = "2" },
+                new Bar() { Value = "3" },
+            };
+        }
+
+        private static int[] CreateInt32Array() => new int[] { 0, 1, 2, 3 };
+
+        [Fact]
+        public static void Fill_Casting()
+        {
+            Bar[] barArray1 = CreateBarArray();
+            Array.Fill<object>(barArray1, new Bar() { Value = "x" });
+            Assert.Equal(Enumerable.Repeat("x", barArray1.Length), barArray1.Select(e => e.Value));
+
+            Bar[] barArray2 = CreateBarArray();
+            Array.Fill<object>(barArray2, new Bar() { Value = "x" }, 1, 2);
+            Assert.Equal(new string[] { "0", "x", "x", "3" }, barArray2.Select(e => e.Value));
+
+            uint[] uintArray1 = (uint[])(object)CreateInt32Array();
+            Array.Fill<uint>(uintArray1, 42);
+            Assert.Equal(Enumerable.Repeat(42, uintArray1.Length), uintArray1);
+
+            uint[] uintArray2 = (uint[])(object)CreateInt32Array();
+            Array.Fill<uint>(uintArray2, 42, 1, 2);
+            Assert.Equal(new int[] { 0, 42, 42, 3 }, uintArray2);
+        }
+
+        [Fact]
+        public static void Fill_Casting_Invalid()
+        {
+            Bar[] barArray1 = CreateBarArray();
+            Assert.Throws<ArrayTypeMismatchException>(() => Array.Fill<object>(barArray1, new Foo()));
+
+            Bar[] barArray2 = CreateBarArray();
+            try
```

Comment:
Don't use try / catch in these tests, as it won't properly handle the failure condition. Instead, use:

```cs
Assert.Throws<ArrayTypeMismatchException>(() => ...);
Assert.Equal(...);
```

These two lines will check that the callback throws, _and_ it will also check that the array has not been mutated, as you are doing here.

See near the top of this file for some other examples of using `Assert.Throws`.